### PR TITLE
functors_task_1_Serhii_Starodumov

### PR DESCRIPTION
**Pull request by SStarodumov**

### CODE REVIEW

File: functors_task_1/student/include/impl.h [hunk 1]
```diff
@@ -8,6 +8,11 @@
  */
 struct BeerOrganizer
 {
+    BeerOrganizer();
+    size_t m_number;
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:12]: _please use more descriptive name_

File: functors_task_1/student/include/impl.h [hunk 2]
```diff
@@ -8,6 +8,11 @@
  */
 struct BeerOrganizer
 {
+    BeerOrganizer();
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:11]: Use in-place initialization:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c45-dont-define-a-default-constructor-that-only-initializes-data-members-use-in-class-member-initializers-instead

File: functors_task_1/student/src/impl.cpp [hunk 3]
```diff
@@ -1,2 +1,37 @@
...
+    }
+    return static_cast<BeerBrand>(m_number++);
+}
+
+BeerOrganizer::BeerOrganizer():m_number(m_min+1) {
+    static_assert(m_max-m_min>1, "You have only None and Max in enum!");
+}
+
+bool isExpensiveBeer(const BeerBrand brand) {
+    constexpr static std::array<BeerBrand, 2> expensiveBeers =
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:17]: I suppose that is overengineering, please simplify with plain logical expression

File: functors_task_1/student/src/impl.cpp [hunk 4]
```diff
@@ -1,2 +1,37 @@
...
+            {BeerBrand::Corona, BeerBrand::HoeGaarden};
+    return std::find(expensiveBeers.cbegin(), expensiveBeers.cend(), brand)!=expensiveBeers.cend();
+}
+
+Cocktail MixingPolicy::mix(const AlcoholDrink alcoholDrink, const NonAlcoholDrink nonAlcoholDrink) {
+    using Ingredients = std::pair<AlcoholDrink, NonAlcoholDrink>;
+    using Recipe = std::pair<Ingredients , Cocktail>;
+    constexpr static std::array<Recipe , 3> recipes = {
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:25]: In such cases, unordered map should be preferred.

But nice try :)

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: functors_task_1/student/src/impl.cpp [hunk 5]
```diff
@@ -1,2 +1,37 @@
...
+}
+
+bool isExpensiveBeer(const BeerBrand brand) {
+    constexpr static std::array<BeerBrand, 2> expensiveBeers =
+            {BeerBrand::Corona, BeerBrand::HoeGaarden};
+    return std::find(expensiveBeers.cbegin(), expensiveBeers.cend(), brand)!=expensiveBeers.cend();
+}
+
+Cocktail MixingPolicy::mix(const AlcoholDrink alcoholDrink, const NonAlcoholDrink nonAlcoholDrink) {
+    using Ingredients = std::pair<AlcoholDrink, NonAlcoholDrink>;
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:23]: +1 for aliases, they make code more readable

File: functors_task_1/student/src/impl.cpp (hunk 4 above)
Comment (SStarodumov) [reply functors_task_1/student/src/impl.cpp]: according to cppref (https://en.cppreference.com/w/cpp/utility/hash) std::pair has no std::hash

Comment (smay1613) [reply functors_task_1/student/src/impl.cpp]: So you should provide your own implementation, it is pretty simple

**Pull request closed without merging**